certbot-apache/certbot_apache/interfaces.py

```diff
@@ -110,35 +113,30 @@ class ParserNode(object):
     it's responsibility of each ParserNode object itself to store its prepending
     whitespaces in order to be able to reconstruct the complete configuration file
     as it was when originally read from the disk.
-    """
 
-    @property
-    @abc.abstractmethod
-    def ancestor(self):  # pragma: no cover
-        """
-        This property contains a reference to ancestor node, or None if the node
-        is the root node of the configuration tree.
+    ParserNode objects should have the following attributes:
 
-        :returns: The ancestor BlockNode object, or None for root node.
-        :rtype: ParserNode
-        """
+    # Reference to ancestor node, or None if the node is the root node of the
+    # configuration tree.
+    ancestor: Optional[ParserNode]
+
+    # True if this node has been modified since last save.
+    dirty: bool
 
-        raise NotImplementedError
+    # Filepath of the file where the configuration element for this ParserNode
+    # object resides.
+    filepath: str
```

Comment:
What should filepath be for the root node?

Reply:
I think the root node should point to the main config which we start parsing from, even though it doesn't have a key name.

Reply:
I made it `Optional[str]` as while if we go with the root node having the file path of the Apache2 root config file, if directives are added on Apache command line, those will not have a filepath.

Reply:
Works for me. Can you document this stuff in the comment for `filepath`?